core/message-tags-3.3.md

```diff
@@ -80,6 +89,24 @@ Individual tag keys MUST only be used a maximum of once per message. Clients
 receiving messages with more than one occurrence of a tag key SHOULD discard all
 but the final occurrence.
 
+### Empty tagged messages
+
+Servers MUST accept `PRIVMSG` and `NOTICE` events sent with client-only tags even if
```

Comment:
> how many places in the znc codebase would this affect?

All modules, including external ones. Because the module callback has been `virtual EModRet OnChanMsg(CNick& Nick, CChan& Channel, CString& sMessage);`. This can't be called for PRIVMSG without text as it would be indistinguishable from PRIVMSG with empty text.

It's not only about ZNC. [KVIrc](https://github.com/kvirc/KVIrc/blob/f8f50bd6a787fdae7ce03cc36f97068399700d66/src/kvirc/kvs/event/KviKvsEventTable.cpp#L1269) would suffer from similar problems.